src/libmongoc/tests/test-mongoc-collection.c

```diff
@@ -5630,6 +5630,62 @@ test_create_indexes_with_opts_commitQuorum_post44 (void *unused)
    mongoc_client_destroy (client);
 }
 
+void
+test_insert_one_reports_id (void)
+{
+   mongoc_client_t *client = test_framework_new_default_client ();
+   mongoc_collection_t *coll = get_test_collection (client, "test_insert_reports_id");
+   bson_error_t error;
+   bool ok;
+
+   // Test inserting one document reports the inserted ID.
+   {
+      bson_t *doc = tmp_bson ("{'_id': 'foo'}");
+      bson_t reply;
+      ok = mongoc_collection_insert_one (coll, doc, NULL /* opts */, &reply, &error);
+      ASSERT_OR_PRINT (ok, error);
+      // Check that `reply` contains the inserted ID in an array `ids`.
+      ASSERT_MATCH (&reply, "{'insertedId': 'foo'}");
+      bson_destroy (&reply);
+   }
+
+   // Test inserting one document reports the generated inserted ID.
+   {
+      bson_t *doc = tmp_bson ("{'foo': 'bar'}");
+      bson_t reply;
+      ok = mongoc_collection_insert_one (coll, doc, NULL /* opts */, &reply, &error);
+      ASSERT_OR_PRINT (ok, error);
+      // Check that `reply` contains the inserted ID in an array `ids`.
+      // Since the driver creates a random ID, only assert it exists.
+      ASSERT_MATCH (&reply, "{'insertedId': {'$exists': true}}");
+      bson_destroy (&reply);
+   }
+
+   // Test inserting one document with a duplicate ID doesn't report the ID.
+   {
+      bson_t *doc1 = tmp_bson ("{'_id': 'baz'}");
+      bson_t reply1;
+      ok = mongoc_collection_insert_one (coll, doc1, NULL /* opts */, &reply1, &error);
+      ASSERT_OR_PRINT (ok, error);
+      // Check that `reply` contains the inserted ID
+      // Since the driver creates a random ID, only assert it exists.
+      ASSERT_MATCH (&reply1, "{'insertedId': 'baz'}");
+
+      // Insert the duplicate document.
+      bson_t *doc2 = tmp_bson ("{'_id': 'baz'}");
+      bson_t reply2;
+      ok = mongoc_collection_insert_one (coll, doc2, NULL /* opts */, &reply2, &error);
+      ASSERT_OR_PRINT (!ok, error);
+      ASSERT_MATCH (&reply2, "{'insertedId': {'$exists': false}}");
+
+      bson_destroy (&reply1);
+      bson_destroy (&reply2);
+   }
+
```

Comment:
Suggest testing insert of a large `_id` (to ensure no leak occurs due to `bson_t` heap allocation). Example:

```c
   // Test inserting one document with a large ID
   {
      // Create a large string of repeating 'A' characters.
      char *large_str = bson_malloc (128);
      memset (large_str, 'A', 128);
      large_str[127] = '\0'; // NULL terminate string.
      bson_t *doc = tmp_bson ("{'_id': '%s'}", large_str);
      bson_t reply;
      ok = mongoc_collection_insert_one (coll, doc, NULL /* opts */, &reply, &error);
      ASSERT_OR_PRINT (ok, error);
      // Check that `reply` contains the inserted ID in an array `ids`.
      ASSERT_MATCH (&reply, "{'insertedId': '%s'}", large_str);
      bson_destroy (&reply);
      bson_free (large_str);
   }
```